### PR TITLE
Apply iftrue and iffalse to tab and allow nested conditions

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because this is a new feature.
Close https://github.com/sonata-project/SonataAdminBundle/issues/5775

## Changelog

```markdown
### Changed
IfTrue and IfFalse apply correctly to tab and with functions
Nested IfTrue and IfFalse works as expected and does not throws exception anymore.
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
I don't think the `IfTrue` and `IfFalse` are documented somewhere.

https://symfony.com/doc/master/bundles/SonataAdminBundle/getting_started/the_form_view.html
https://symfony.com/doc/master/bundles/SonataAdminBundle/reference/action_show.html

**Comment by VincentLanglet** [comment]
@phansys I added tests, is it what you wanted ?

**Comment by phansys** [comment]
> I don't think the `IfTrue` and `IfFalse` are documented somewhere @greg0ire.
> 
> https://symfony.com/doc/master/bundles/SonataAdminBundle/getting_started/the_form_view.html
> https://symfony.com/doc/master/bundles/SonataAdminBundle/reference/action_show.html

Please, see [Fields Definition](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/architecture.rst#fields-definition) by instance.

**Comment by phansys** [comment]
> @phansys I added tests, is it what you wanted ?

Yes!
We should check the nested behavior too. IIRC, it will throw an exception.

**Comment by VincentLanglet** [comment]
> Please, see [Fields Definition](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/reference/architecture.rst#fields-definition) by instance.

This does give an example. I can add an example with both `with`  and `ifTrue` but it seems weird since there is no use of the `tab` or `with` function on this page.

When I first discovered theses functions, I was expecting to use them with `tab` and `with` function.So I think this will be more intuitive with this PR.

**Comment by VincentLanglet** [comment]
> We should check the nested behavior too. IIRC, it will throw an exception.

Oh, I didn't thought about nested `IfTrue`. Is there a reason, we didn't allow
```
$formMapper
  ->IfTrue(a)
    ->add('foo')
    ->IfTrue(b)
      ->add('bar')
    ->ifEnd()
  ->ifEnd()
```

**Comment by phansys** [comment]
I guess that is a technical limitation, but I'm not sure.

**Comment by VincentLanglet** [comment]
@phansys This https://github.com/sonata-project/SonataAdminBundle/pull/5779/commits/3a802242b706b1c491752c83c518680d8e4c3b66 could allow nested IfTrue/IfFalse.

Since IfTrue/IfFalse apply to `tab` and `with` with this PR, I think we should allow nested conditions.
```
$mapper
  ->IfTrue(a)
    ->tab(b)
      ->IfFalse(c)
        ->with(d)
          ->add(e)
          ->IfTrue(f) 
            ->add(g)
          ->IfEnd()
          ->add(h)
        ->end()
      ->IfEnd()
      ->with(i)
        ->add(j)
      ->end()
    ->end()
  ->IfEnd()
  ->tab(k)
    ->with(l)
      ->add(m)
    ->end()
  ->end()
```
Could be a use case.

**Comment by VincentLanglet** [comment]
@phansys You're the one requesting change, is it ok for you now ? :)

**Comment by phansys** [comment]
> @phansys You're the one requesting change, is it ok for you now ? :)

I'd like to test it locally. I'll try to check this today.
Thank you so much!

**Comment by greg0ire** [comment]
@VincentLanglet if you want to help @phansys , you can use this script: https://gist.github.com/greg0ire/a404831add1247d2bc20fa11107c5d5e

It will produce instructions to test your PR.

File: src/Mapper/BaseGroupedMapper.php [hunk 1]
```diff
@@ -306,4 +302,20 @@ protected function getCurrentGroupName()
 
         return $this->currentGroup;
     }
+
+    /**
+     * Check if all apply conditions are respected.
+     *
+     * @return bool
+     */
+    protected function shouldApply()
```
Comment (franmomu) [inline src/Mapper/BaseGroupedMapper.php]: You can add return type

File: tests/Form/FormMapperTest.php [hunk 2]
```diff
@@ -300,56 +300,56 @@ public function testIfFalseCombination(): void
 
     public function testIfTrueNested(): void
     {
-        $this->expectException(\LogicException::class);
-        $this->expectExceptionMessage('Cannot nest ifTrue or ifFalse call');
-
         $this->formMapper->ifTrue(true);
         $this->formMapper->ifTrue(true);
+        $this->formMapper->add('fooName');
```
Comment (franmomu) [inline tests/Form/FormMapperTest.php]: Shouldn't this (and the other tests) be more like:
```php
$this->formMapper
    ->ifTrue(true)
        ->ifTrue(true)
            ->add('fooName')
        ->ifEnd()
    ->ifEnd()
;
```

IMHO it's more understandable

File: src/Mapper/BaseGroupedMapper.php [hunk 3]
```diff
@@ -179,17 +177,11 @@ public function ifTrue($bool)
...
      * @return $this
      */
     public function ifFalse($bool)
     {
-        if (null !== $this->apply) {
-            throw new \LogicException('Cannot nest ifTrue or ifFalse call');
-        }
-
-        $this->apply = (false === $bool);
+        array_push($this->apply, (false === $bool));
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
        $this->apply[] = false === $bool;
```

File: src/Mapper/BaseGroupedMapper.php [hunk 4]
```diff
@@ -159,17 +163,11 @@ public function with($name, array $options = [])
...
      * @return $this
      */
     public function ifTrue($bool)
     {
-        if (null !== $this->apply) {
-            throw new \LogicException('Cannot nest ifTrue or ifFalse call');
-        }
-
-        $this->apply = (true === $bool);
+        array_push($this->apply, (true === $bool));
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
        $this->apply[] = true === $bool;
```

**Comment by phansys** [comment]
@VincentLanglet, I've tested these changes locally and they seem to behave as expected :+1:

**Comment by VincentLanglet** [comment]
@OskarStark Yes, it change the existing behaviour. Before, the use of nested conditions returned an exception. Now it uses all the conditions to display or not a field.

I don't consider this like a breaking change, more like a new feature.

File: src/Mapper/BaseGroupedMapper.php [hunk 5]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
 
         return $this->currentGroup;
     }
+
+    /**
+     * Check if all apply conditions are respected.
+     */
+    protected function shouldApply(): bool
```
Comment (core23) [inline src/Mapper/BaseGroupedMapper.php]: Can you please make this final?

File: src/Mapper/BaseGroupedMapper.php [hunk 6]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+    {
+        foreach ($this->apply as $condition) {
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
        return !\in_array(false, $apply, true);
```

File: src/Mapper/BaseGroupedMapper.php [hunk 7]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+            if (!$condition) {
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
```

File: src/Mapper/BaseGroupedMapper.php [hunk 8]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+                return false;
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
```

File: src/Mapper/BaseGroupedMapper.php [hunk 9]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+            }
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
```

File: src/Mapper/BaseGroupedMapper.php [hunk 10]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+        }
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
```

File: src/Mapper/BaseGroupedMapper.php [hunk 11]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
```

File: src/Mapper/BaseGroupedMapper.php [hunk 12]
```diff
@@ -306,4 +302,18 @@ protected function getCurrentGroupName()
...
+        return true;
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
```

File: src/Mapper/BaseGroupedMapper.php [hunk 13]
```diff
@@ -33,9 +33,9 @@ abstract class BaseGroupedMapper extends BaseMapper
     protected $currentTab;
 
     /**
-     * @var bool|null
+     * @var array
```
Comment (phansys) [inline src/Mapper/BaseGroupedMapper.php]:
```suggestion
     * @var bool[]
```

**Comment by VincentLanglet** [comment]
@core23 @phansys done !

**Comment by VincentLanglet** [comment]
@core23 I don't understand exactly what you want.
If an `ifEnd` is missing there will be no error.

```
public function testIfTrueNestedWithoutIfEnd(): void
{
	$this->formMapper->ifTrue(true)->ifTrue(true)->add('fooName');
        $this->assertTrue($this->formMapper->has('fooName'));
}
```

**Comment by core23** [comment]
> @core23 I don't understand exactly what you want.
> If an `ifEnd` is missing there will be no error.

As said, this is not related to your change, but I would expect an error, if you give a wrong or incomplete defintion. Can we add a deprecation warning there?

**Comment by VincentLanglet** [comment]
I can add an error if you call ifEnd without something to end.
But I don't see how to add an error if you forget to call ifEnd.

**Comment by OskarStark** [comment]
Thanks, great feature 😃

**Comment by core23** [comment]
> I can add an error if you call ifEnd without something to end.
> But I don't see how to add an error if you forget to call ifEnd.

I created a new issue for this: https://github.com/sonata-project/SonataAdminBundle/issues/5806

**Pull request merged**